description: use system jpeg library
author: Michael Gilbert <mgilbert@debian.org>

--- a/src/BUILD.gn
+++ b/src/BUILD.gn
@@ -752,7 +752,6 @@ group("gn_all") {
       "//third_party/angle:libEGL",
       "//third_party/angle:libGLESv2",
       "//third_party/leveldatabase:leveldb_test_targets",
-      "//third_party/libjpeg_turbo:simd",
       "//third_party/opus:opus_compare",
       "//third_party/opus:opus_demo",
       "//third_party/opus:test_opus_api",
@@ -767,7 +766,6 @@ group("gn_all") {
       if (!is_android) {
         deps += [ "//chrome/test:load_library_perf_tests" ]
       }
-      deps += [ "//third_party/libjpeg_turbo:simd_asm" ]
     }
     if ((is_linux || is_chromeos) && current_toolchain == host_toolchain) {
       deps += [ "//v8:v8_shell" ]
