description: disable dependency on chrome/android
author: Michael Gilbert <mgilbert@debian.org>

--- a/src/BUILD.gn
+++ b/src/BUILD.gn
@@ -346,7 +346,7 @@ group("gn_all") {
   # Allow building md5sum tool for other OSes (android is added below).
   # For Windows, see: crbug.com/1090428.
   # TODO(cavalcantii): move it out of 'android' folder.
-  if (!is_win) {
+  if (is_android) {
     deps += [ "//tools/android/md5sum" ]
   }
 
