description: additional swiftshader stuff exposed by official builds
author: Andres Salomon <dilinger@debian.org>

--- a/src/chrome/BUILD.gn
+++ b/src/chrome/BUILD.gn
@@ -1719,7 +1719,6 @@ if (is_linux || is_chromeos) {
         ":chrome_symbols",
       ]
       if (is_linux) {
-        deps += [ ":swiftshader_vk_symbols" ]
       }
       if (!is_chromeos) {
         deps += [ ":angle_libvulkan_symbols" ]
@@ -1752,6 +1751,7 @@ if (is_linux || is_chromeos) {
 
       deps = [ "//components/crash/core/app:chrome_crashpad_handler" ]
     }
+if (false) {
     extract_symbols("swiftshader_vk_symbols") {
       binary = "$root_out_dir/libvk_swiftshader.so"
 
@@ -1764,6 +1764,7 @@ if (is_linux || is_chromeos) {
 
       deps = [ "//third_party/swiftshader/src/Vulkan:swiftshader_libvulkan" ]
     }
+}
     extract_symbols("angle_egl_symbols") {
       binary = "$root_out_dir/libEGL.so"
 
