From 829825f0609d20d280a36d653376e322a30486d9 Mon Sep 17 00:00:00 2001
Date: Tue, 1 Mar 2022 08:19:19 +0800
Subject: [PATCH] Reset all planes state due to hotplug dirty state.

To fix the overlay state hang when surface changed the position
after hotplug.

---
 libweston/backend-drm/drm.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 2df80157..f8e3b702 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -2285,6 +2285,15 @@ drm_backend_update_heads(struct drm_backend *b, struct udev_device *drm_device)
 		drm_head_destroy(head);
 	}
 
+        /* Reset all planes state due to dirty status after hotplug. */
+        struct drm_plane *plane;
+
+        wl_list_for_each(plane, &b->plane_list, link) {
+               if (plane->type != WDRM_PLANE_TYPE_OVERLAY)
+                       continue;
+               drm_plane_reset_state(plane);
+        }
+
 	drm_backend_update_unused_outputs(b, resources);
 
 	drmModeFreeResources(resources);
-- 
2.17.1

